pulsar-broker/src/main/java/org/apache/pulsar/compaction/StrategicTwoPhaseCompactor.java

```diff
@@ -385,6 +385,7 @@ private <T> void phaseTwoLoop(String topic, Iterator<Message<T>> reader,
                             promise.completeExceptionally(e);
                             return;
                         }
+                        outstanding.release(MAX_OUTSTANDING);
```

Comment:
should we move this after `promise.complete(null);` ?

Reply:
No, I found a deadlock if we don't release this semaphore before completing the future.

deadLock example

```
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Semaphore;
import org.junit.Test;

public class SemaphoreTest {
    
    @Test
    public void deadlockTest()  {
        Semaphore sm = new Semaphore(2);

        ExecutorService executor = Executors.newSingleThreadExecutor();
        CompletableFuture<String> future = new CompletableFuture();
        CompletableFuture.supplyAsync(() -> {
            System.out.println("starting");
            try {
                sm.acquire(2);
                System.out.println("acquired 2");
            } catch (InterruptedException e) {
            }
            try {
                Thread.sleep(1000*5);
            } catch (InterruptedException e) {
            }
            System.out.println("return!");
            future.complete(""); // hangs here.
            System.out.println("proceeding");
            sm.release(2);
            System.out.println("released 2");
            return "";
        }, executor);
        future.thenCompose(x -> {
            try {
                System.out.println("acquiring one!");
                sm.acquire(); // hangs here.
                System.out.println("acquired 1");
                sm.release();
            } catch (InterruptedException e) {
            }
            return CompletableFuture.completedFuture("");
        }).join();
    }
}
```